clang/lib/CIR/CodeGen/CIRGenModule.cpp

```diff
@@ -3525,8 +3533,77 @@ void CIRGenModule::buildGlobalAnnotations() {
   deferredAnnotations.clear();
 }
 
+mlir::LLVM::TBAATypeDescriptorAttr CIRGenModule::getTBAATypeInfo(QualType QTy) {
```

Comment:
We don't want to use LLVM IR Dialect specific stuff, like `mlir::LLVM::TBAATypeDescriptorAttr` directly in CIRGen or anywhere before LowerToLLVM.

The idea is to design our CIR version of this, even if very similar. I wonder if we really need to pass in the same amount of information LLVM requires though, because we have the types until lowering time, perhaps we could do something a bit more terse and expand to the full metadata content during LLVM lowering? For example, we don't need to encode a string for `id = "omnipotent char"` while in CIR, that could be just a enum kind if we need to encode it early (do we tho?). The overall idea is to infer things that we can during lowering while only encoding information during CIRGen that it's hard to do later on. Note that CIR can be lowered to non-LLVM targets, so we need to be a bit strategical here.

Also, the tbaa information should be first class into cir.load/cir.store, so `{tbaa = [#tbaa_tag]}` should probably just be `tbaa([#tbaa_tag])` or something that's not only dumped into the generic attributes space.

This is a big PR that touches lots of things, I think we should slightly break this up so we can land the bulk faster and spend time on a PR for the specific attribute design. My suggestion is that first PR should just add the overall skeleton for CIRGen, this effectively means changing the PR such that:

1. Create a dummy `mlir::cir::TBAA_Attr`, shouldn't have anything in it.
2. Replace all LLVM specific attribute bits like `mlir::LLVM::TBAATypeDescriptorAttr`, to `mlir::cir::TBAA_Attr`.
3. The current implementation of everything that returns `mlir::cir::TBAA_Attr` should create an empty `mlir::cir::TBAA_Attr`.
4. Attach to loads/stores just like you are currently doing.
5. One new really simple CIR test that checks the presence of `mlir::cir::TBAA_Attr` on cir.load/cir.store.
6. No LLVM lowering.

Reply:
Thanks for your comments. I completely agree with your points about not using LLVM-specific elements in CIRGen. It's important to keep things clean for potential non-LLVM targets.

I will break down the PR as you suggested.